File: Arrays and Hashing/group-anagrams.py
from collections import defaultdict


class Solution:
    def groupAnagrams(self, strs):
        hash_map = defaultdict(list)
        for str in strs:
            count = [0] * 26
            for char in str:
                index = ord(char) - ord("a")
                count[index] += 1
            hash_map[tuple(count)].append(str)
        return hash_map.values()

File: Arrays and Hashing/group-anagrams.js
// time O(mn) space O(n)
var groupAnagrams = function (strs) {
  let hashMap = new Map();
  for (let str of strs) {
    let count = Array(26).fill(0);
    for (let char of str) {
      let index = char.charCodeAt(0) - "a".charCodeAt(0);
      count[index]++;
    }
    const key = JSON.stringify(count);
    let prevValue = hashMap.get(key);
    if (prevValue) {
      prevValue.push(str);
      hashMap.set(key, prevValue);
    } else {
      hashMap.set(key, [str]);
    }
  }

  return [...hashMap.values()];
};
